Transition rolled back tasks into `Completed(name="RolledBack")`

collincchoy opened the pull request:
This PR is a successor to #14662 which previously had similar changes but was reduced in scope. It adds:

When using task.on_rollback hooks, when the task is rolled back, this PR causes the task to update its state to a Completed(name="RolledBack") state rather than a plain Completed state.

## Example
For example, this flow using transactions contains a task that completes successfully but is then rolled back due to another part of the transaction failing. That task should now be in a Completed(name="RolledBack") state.

```python
import os
import typing

from prefect import flow, task
from prefect.transactions import Transaction, transaction

@task
def write_file(contents: str):
    "Writes to a file."
    with open("side-effect.txt", "w") as f:
        f.write(contents)


@write_file.on_rollback
def del_file(transaction: Transaction):
    "Deletes file."
    os.unlink("side-effect.txt")


@task
def quality_test():
    "Checks contents of file."
    with open("side-effect.txt", "r") as f:
        data = f.readlines()

    if len(data) < 2:
        raise ValueError("Not enough data!")


@flow
def pipeline(contents: str):
    with transaction():
        write_file(contents)
        quality_test()
```

![image](https://github.com/user-attachments/assets/1361f04b-6a97-4aa1-a4f0-ec55903b30fd)

Closes #14581

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [x] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [x] If this pull request adds functions or classes, it includes helpful docstrings.

collincchoy commented on src/prefect/task_engine.py:
```diff
@@ -716,6 +725,22 @@ def start(
         result_factory = getattr(TaskRunContext.get(), "result_factory", None)
 
+        def __prefect_set_task_state_to_rolled_back(txn: Transaction):
+            if txn.state != TransactionState.STAGED:
```
See prior discussion [here](https://github.com/PrefectHQ/prefect/pull/14662#discussion_r1688712832).

This short-circuit is primarily so that a task like:
```python
@task
def i_fail():
    raise ValueError('whoops')
```
doesn't now transition through a Completed(name="RolledBack") state before retrying and/or failing. The task should only be in this state if it had previously completed, but was then rolled back and its rollback hooks run.

collincchoy commented on tests/test_task_engine.py:
```diff
@@ -2276,3 +2277,114 @@ def foo():
+
+
+class TestTransactionHooks:
+    async def test_task_transitions_to_rolled_back_on_transaction_rollback(
```
This is the only new test - the rest were just moved over from test_tasks.py

collincchoy commented on src/prefect/task_engine.py:
```diff
@@ -506,6 +516,18 @@ def record_terminal_state_timing(self, state: State) -> None:
                     )
 
+    def handle_rollback(self, txn: Transaction) -> None:
+        # transaction rollbacks can occur outside of the engine's context
+        # so we need to ensure a client is available
+        with ClientContext.get_or_create() as client_ctx:
```
With client-side task orchestration, we don't need this, but making it work with or w/o keeps the test suite happier.

collincchoy commented on tests/test_tasks.py:
```diff
@@ -5104,83 +5104,3 @@ def add_em_up(*args, **kwargs):
         assert await get_background_task_run_parameters(
             add_em_up, future.state.state_details.task_parameters_id
         ) == {"parameters": {"args": (42,), "kwargs": {"y": 42}}, "context": ANY}
-
-
-def test_rollback_errors_are_logged(caplog):
```
this entire block was just cut & pasted into test_task_engine.py::TestTransactionHooks. 
no changes other than adding `self` args.

cicdw reviewed (changes requested):
I'm afraid this still has the same issues as before in that it will cause every transaction to have a task run engine associated with it, and it will engage that engine after it's already been closed.

While it's possible this implementation could pass every test right now, because the engine is now tightly coupled to transactions it's one seemingly innocent refactor away from breaking (because it requires knowing that this one method might get called after engine shutdown).

collincchoy commented:
> I'm afraid this still has the same issues as before in that it will cause every transaction to have a task run engine associated with it, and it will engage that engine after it's already been closed.
> 
> While it's possible this implementation could pass every test right now, because the engine is now tightly coupled to transactions it's one seemingly innocent refactor away from breaking (because it requires knowing that this one method might get called after engine shutdown).

Hm okay. I tried addressing concerns from the prior PR's change that allowed access to an engine's client outside of its run context by instead creating and injecting my own client, but it sounds like the primary concern here is memory consumption with a transaction scoped over many tasks where this adds a reference to the task engine into the task run's transaction? 

and/or is the concern that this implementation is brittle for if other instance methods intended to run within the engine's run context were to be added into `set_state`? 

It sounds like you're looking for just:
```python
        emit_task_run_state_change_event(
            task_run=self.task_run,
            initial_state=last_state,
            validated_state=self.task_run.state,
            follows=self._last_event,
        )
```

inside this rollback handler (and not as an instance method)?

I was assuming that we'd want to invoke other task state transitioning logic in the engine's `set_state` here even w/ client-side-only task orchestration stuff, but is that false?

cicdw commented:
@collincchoy yea my primary concern is the design aspect - the engine manages lifecycles for its dependencies (a client, a transaction, a run context, etc.) which is why it has a start/stop lifecycle as well that locks down its interface. Explicitly avoiding that lifecycle in this one instance means the lifecycle can't be depended upon anymore, and the special case isn't transparent because it happens away from the code by an object that calls this object (within the transaction code).  So for example, if someone introduced a reliance on run context in `set_state`, they'd have to account for the fact that the run context could change in the special case of a rollback (which they'd have know way of knowing unless they knew the codebase really well).

That was a long-winded way of saying yea, a basic event emission should be sufficient here since it's only a name change (and not a true state update).  I think you could use your same hook-wrapper approach and make a function that encapsulates the run information as a rollback hook.  

**cc:** @desertaxle in case you had any memory concerns

cicdw reviewed (approved):
This approach does still couple the state change to the task engine despite it being decoupled from the lifecycle of the engine, which I think could result in unexpected bugs in the future. 

That being said, the new state change machinery does require the `follows` kwarg which requires keeping some reference to the last state, so I think we should merge this and @desertaxle and I can monitor for opportunities to refactor it as transitions gain adoption.
